File: P1314 Matrix Block Sum/solution.py
class Solution:
    # Time & space: O(m * n)
    def matrixBlockSum(self, mat: List[List[int]], K: int) -> List[List[int]]:
        m, n = len(mat), len(mat[0])
        sums = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
            for j in range(n):
                sums[i+1][j+1] = sums[i + 1][j] + sums[i][j + 1] - sums[i][j] + mat[i][j]
        # result = [[0] * n] * m => wrong. it creates m references of [[0] * n]
        result = [[0] * n for _ in range(m)]
        for i in range(m):
            for j in range(n):
                r1, c1, r2, c2 = max(0, i - K), max(0, j - K), min(m, i + K + 1), min(n, j + K + 1)
                result[i][j] = sums[r2][c2] - sums[r2][c1] - sums[r1][c2] + sums[r1][c1]
        return result

File: P1314 Matrix Block Sum/Solution.java
class Solution {
    public int[][] matrixBlockSum(int[][] mat, int K) {
        int m = mat.length;
        int n = mat[0].length;
        int[][] sum = new int[m+1][n+1];
         for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                sum[i+1][j+1] = sum[i + 1][j] + sum[i][j + 1] - sum[i][j] + mat[i][j];
            }
        }
        int[][] result = new int[m][n];
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                int r1 = Math.max(0, i - K), c1 = Math.max(0, j - K);
                int r2 = Math.min(m, i + K + 1), c2 = Math.min(n, j + K + 1);
                result[i][j] = sum[r2][c2] - sum[r2][c1] - sum[r1][c2] + sum[r1][c1];
            }
        }
        return result;
    }
}